pkg/tfbridge/auto_aliasing.go

```diff
@@ -95,20 +95,75 @@ func (info *ProviderInfo) MustApplyAutoAliases() {
 // - Edit [SchemaInfo.MaxItemsOne]
 //
 // The goal is to always maximize backwards compatibility and reduce breaking changes for
-// the users of the Pulumi providers.
+// the users of the Pulumi providers. The basic functionality behind each action is
+// identical; ApplyAutoAliases keeps a record of which TF token maps to which Pulumi
+// token, which fields have MaxItemsOne (true or false), and what version the record is
+// from.
 //
-// Resource aliases help mask TF provider resource renames or changes in mapped tokens so
-// older programs continue to work.  See [ResourceInfo.RenameResourceWithAlias] and
-// [ResourceInfo.Aliases] for more details.
+// For example, this is the (abbreviated & modified) history for GCP's compute autoscalar:
 //
-// [SchemaInfo.MaxItemsOne] changes are also important because they involve flattening and
-// pluralizing names. Collections (lists or sets) marked with MaxItems=1 are projected as
-// scalar types in Pulumi SDKs. Therefore changes to the MaxItems property may be breaking
-// the compilation of programs as the type changes from `T to List[T]` or vice versa. To
-// avoid these breaking changes, this method undoes any upstream changes to MaxItems using
-// [SchemaInfo.MaxItemsOne] overrides. This happens until a major version change is
-// detected, and then overrides are cleared. Effectively this makes sure that upstream
-// MaxItems changes are deferred until the next major version.
+//	"google_compute_autoscaler": {
+//	    "current": "gcp:compute/autoscalar:Autoscalar",
+//	    "past": [
+//	        {
+//	            "name": "gcp:auto/autoscalar:Autoscalar",
+//	            "inCodegen": true,
+//	            "majorVersion": 6
+//	        },
+//	        {
+//	            "name": "gcp:auto/scaler:Scaler",
+//	            "inCodegen": false,
+//	            "majorVersion": 5
+//	        }
+//	    ],
+//	    "majorVersion": 6,
+//	    "fields": {
+//	        "autoscaling_policy": {
+//	            "maxItemsOne": true,
+//	            "elem": {
+//	                "fields": {
+//	                    "cpu_utilization": {
+//	                        "maxItemsOne": false
+//	                    }
+//	                }
+//	            }
+//	        }
+//	    }
+//	}
+//
+// I will address each action as it applies to `"google_compute_autoscaler" in turn:
+//
+// # Call [ProviderInfo.RenameResourceWithAlias] or [ProviderInfo.RenameDataSource]
+//
+// "google_compute_autoscaler.majorVersion" tells us that this record was last updated at
+// major version 6. One of the previous names is also at major version 6, so we want to
+// keep full backwards compatibility.
+//
+// ApplyAutoAliases will call [ProviderInfo.RenameResourceWithAlias] to create a SDK entry
+// for the old Pulumi token ("gcp:auto/autoscalar:Autoscalar"). This token will no longer
+// be created when `make tfgen` is run on version 7.
```

Comment:
Is this because we don't have an entry for v7? because v7 is getting rid of this resource upstream?

Reply:
Clarified in https://github.com/pulumi/pulumi-terraform-bridge/pull/1416/commits/f29d643740970b44f18a118e4e9e3d02d1aaa274.